letsencrypt/models/letsencrypt.py

```diff
@@ -0,0 +1,157 @@
+        account_key = os.path.join(data_dir, 'account.key')
+        if not os.path.isfile(account_key):
+            _logger.info('generating rsa account key')
+            self.call_cmdline([
+                'openssl', 'genrsa', '-out', account_key,
+                str(DEFAULT_KEY_LENGTH),
+            ])
+        assert os.path.isfile(account_key), 'failed to create rsa key'
+        return account_key
+
+    @api.model
+    def generate_domain_key(self, domain):
+        domain_key = os.path.join(self.get_data_dir(), '%s.key' % domain)
+        if not os.path.isfile(domain_key):
+            _logger.info('generating rsa domain key for %s', domain)
+            self.call_cmdline([
+                'openssl', 'genrsa', '-out', domain_key,
+                str(DEFAULT_KEY_LENGTH),
+            ])
+        return domain_key
+
+    @api.model
+    def validate_domain(self, domain):
+        if domain in ['localhost', '127.0.0.1'] or\
+                domain.startswith('10.') or\
+                domain.startswith('172.16.') or\
+                domain.startswith('192.168.'):
+            raise exceptions.Warning(
+                _("Let's encrypt doesn't work with private addresses!"))
+
+    @api.model
+    def generate_csr(self, domain):
+        domains = [domain]
+        i = 0
+        while self.env['ir.config_parameter'].get_param(
+                'letsencrypt.altname.%d' % i):
+            domains.append(
+                self.env['ir.config_parameter']
+                .get_param('letsencrypt.altname.%d' % i)
+            )
+            i += 1
+        _logger.info('generating csr for %s', domain)
+        if len(domains) > 1:
+            _logger.info('with alternative subjects %s', ','.join(domains[1:]))
+        config = self.env['ir.config_parameter'].get_param(
+            'letsencrypt.openssl.cnf', '/etc/ssl/openssl.cnf')
+        csr = os.path.join(self.get_data_dir(), '%s.csr' % domain)
+        with tempfile.NamedTemporaryFile() as cfg:
+            cfg.write(open(config).read())
+            if len(domains) > 1:
+                cfg.write(
+                    '\n[SAN]\nsubjectAltName=' +
+                    ','.join(map(lambda x: 'DNS:%s' % x, domains)) + '\n')
+            cfg.file.flush()
+            cmdline = [
+                'openssl', 'req', '-new',
+                self.env['ir.config_parameter'].get_param(
+                    'letsencrypt.openssl.digest', '-sha256'),
+                '-key', self.generate_domain_key(domain),
+                '-subj', '/CN=%s' % domain, '-config', cfg.name,
+                '-out', csr,
+            ]
+            if len(domains) > 1:
+                cmdline.extend([
+                    '-reqexts', 'SAN',
+                ])
+            self.call_cmdline(cmdline)
+        return csr
+
+    @api.model
+    def cron(self):
+        domain = urlparse.urlparse(
+            self.env['ir.config_parameter'].get_param(
+                'web.base.url', 'localhost')).netloc
```

Comment:
If the `web.base.url` equals to `http://my.domain:8069`, the validation process produces this error:

```
Traceback (most recent call last):
  File "/home/odoo/odoo/server/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/odoo/odoo/server/openerp/http.py", line 574, in dispatch
    result = self._call_function(**self.params)
  File "/home/odoo/odoo/server/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/odoo/odoo/server/openerp/service/model.py", line 118, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/odoo/odoo/server/openerp/http.py", line 307, in checked_call
    return self.endpoint(*a, **kw)
  File "/home/odoo/odoo/server/openerp/http.py", line 803, in __call__
    return self.method(*args, **kw)
  File "/home/odoo/odoo/server/openerp/http.py", line 403, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/odoo/server/addons/web/controllers/main.py", line 948, in call_button
    action = self._call_kw(model, method, args, {})
  File "/home/odoo/odoo/server/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File "/home/odoo/odoo/server/openerp/api.py", line 268, in wrapper
    return old_api(self, *args, **kwargs)
  File "/home/odoo/odoo/server/openerp/api.py", line 439, in old_api
    result = new_api(recs, *args, **kwargs)
  File "/home/odoo/odoo/server/openerp/api.py", line 443, in new_api
    result = [method(rec, *args, **kwargs) for rec in self]
  File "/home/odoo/odoo/addons_oca_server_tools/cron_run_manually/ir_cron.py", line 72, in run_manually
    return method(*args)
  File "/home/odoo/odoo/server/openerp/api.py", line 266, in wrapper
    return new_api(self, *args, **kwargs)
  File "/home/odoo/odoo/addons_oca_server_tools/letsencrypt/models/letsencrypt.py", line 141, in cron
    account_key, csr, acme_challenge, log=_logger, CA=DEFAULT_CA)
  File "/home/odoo/odoo/addons_oca_server_tools/letsencrypt/models/acme_tiny.py", line 105, in get_crt
    raise ValueError("Error requesting challenges: {0} {1}".format(code, result))
ValueError: Error requesting challenges: 400 {"type":"urn:acme:error:malformed","detail":"Error creating new authz :: Invalid character in DNS name","status":400}
```

Is it related to the presence of `:`? If so, maybe throws an explicit error message if the `ParseResult(...).netloc` != `ParseResult(...).hostname` (I don't know if it is the right fix) and invite the administrator to run on ports 80/443 with a correct `web.base.url`.